feat(checkbox-table): Add checkbox column and selection handling

dt-mafe opened the pull request:
### <strong>Pull Request</strong>

<hr>

Early and rough draft for a checkbox column and respective selection handling in the table data source

#### Type of PR

<!-- Bugfix (non-breaking change which fixes an issue) -->
Feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or change that would cause existing functionality to not work as expected) -->
<!-- Documentation update (changes to documentation) -->
<!-- Other (if none of the above apply) -->

#### Checklist

- [x] I have read the CONTRIBUTING doc and I follow the PR guidelines
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

ffriedl89 commented:
Talked to @dt-mafe and this is work in progress right now.

lukasholzer commented:
> Talked to @dt-mafe and this is work in progress right now.

@dt-mafe if it is work in progress would you be so kind to set the PR in the draft state?

![image](https://user-images.githubusercontent.com/11156362/86561041-f74c6a00-bf5f-11ea-87a2-e227ac2c3ca2.png)

dt-mafe commented:
I am sorry about not setting this PR back into draft state after the initial feedback, but this pull-request should be ready for review now.

tomheller commented on libs/barista-components/table/src/expandable/expandable-row.ts:
```diff
@@ -170,6 +170,7 @@ export class DtExpandableRow extends DtRow
     elementRef: ElementRef,
   ) {
     super(elementRef);
+    this._table._hasExpandableRows = true;
```
@ffriedl89 How is this going to work combined with the `when` condition on rows? 
If there are expandable rows should initially, and the expression driving the `when` condition changes resulting in no expandable rows shown, how does that one reset on the table?

dt-mafe replied:
The only reason this property seems to exist at the moment is for styling reasons since expandable rows use a 2px padding vs the 1px padding of the default rows which messes up the alignment of the header-checkbox with those in the table rows.
Note that this problem is not yet fixed and you would have to manually adjust styling when switching between expandable/none expandable rows in that situation to ensure perfect alignment of the header checkbox with the table. 

We could avoid this altogether if left/right paddings were consistent between expandable and none-expandable rows which would be possible since the expandable rows only really need the additional pixel on the top and bottom for the 2px thick border used there in the expanded state. Not sure if that could cause any other styling issues due top/bottom padding no longer being the same as left/right padding.

tomheller commented on libs/barista-components/table/src/selection/selector.spec.ts:
```diff
@@ -0,0 +1,328 @@
+  describe('without limit', () => {
+    beforeEach(() => {
+      TestBed.compileComponents();
+      fixture = TestBed.createComponent(DtTableSelectionComponentForTesting);
+      component = fixture.componentInstance;
+      fixture.detectChanges();
```
Nitpick:
You could use the provided helper function for this. It creates the component and calls the first detect changes.
Add the import
```
import { createComponent } from '@dynatrace/testing/browser';
```

```suggestion
      fixture = createComponent(DtTableSelectionComponentForTesting);
      component = fixture.componentInstance;
```

tomheller commented on libs/barista-components/table/src/selection/selectors.ts:
```diff
@@ -0,0 +1,184 @@
+  /** The rows data used for selection */
+  @Input() row: T;
+
+  @Input() ariaLabel: '';
```
We try to stay close to native code as long as possible, please rename the input to `aria-label` and do not assign a value. Angular will not render the attribute if it is undefined, which is more helpful than an empty aria-label string. 
```
  /** Aria label of the row selector checkbox. */
  @Input('aria-label') ariaLabel: string;
```

tomheller commented on libs/barista-components/table/src/selection/selectors.ts:
```diff
@@ -0,0 +1,184 @@
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class DtTableHeaderSelector<T> implements OnDestroy {
+  @Input() ariaLabel: '';
```
Same here, please rename the input to `aria-label` and do not assign a default value empty string.

tomheller commented on libs/barista-components/table/src/table-data-source.ts:
```diff
@@ -374,6 +373,61 @@ export class DtTableDataSource<T> extends DataSource<T> {
+  //   if (limit) {
+  //     const currentSelection = this.selectionModel.selected;
+  //     const addSelection = selectableData
+  //       .filter((entry) => !currentSelection.includes(entry))
+  //       .slice(0, limit - currentSelection.length);
+  //     return addSelection.concat(currentSelection);
+  //   } else {
+  //     return selectableData;
+  //   }
+  // }
+
+  // private _updateGlobalSelectionStates(selectableData: T[]): void {
+  //   if (isNil(this._selection)) {
+  //     return;
+  //   }
+  //   this._selection.allSelected =
+  //     (this.selectionModel.selected.length === selectableData.length ||
+  //       (this._selection.selectionLimit !== undefined &&
+  //         this.selectionModel.selected.length >=
+  //           this._selection.selectionLimit)) &&
+  //     this.selectionModel.selected.length > 0;
+  //   this._selection.anySelected =
+  //     this.selectionModel.selected.length !== selectableData.length &&
+  //     this.selectionModel.selected.length > 0;
+  // }
+
+  // private _isSelectable = (data: T) => {
+  //   return (
+  //     this.selectionModel.isSelected(data) ||
+  //     (!isNil(this._selection) &&
+  //       this._selection.selectable(data) &&
+  //       (this._selection.selectionLimit === undefined ||
+  //         this.selectionModel.selected.length < this._selection.selectionLimit))
+  //   );
+  // };
+
+  // private _getSelectableData(data: T[]): T[] {
+  //   return data.filter(this._isSelectable);
+  // }
+
```
Please remove commented code.

tomheller commented on libs/barista-components/table/src/table-data-source.ts:
```diff
@@ -112,6 +110,7 @@ export class DtTableDataSource<T> extends DataSource<T> {
    */
   private _renderChangesSubscription = Subscription.EMPTY;
   private _searchChangeSubscription = Subscription.EMPTY;
+  private _selectionChangeSubscription = Subscription.EMPTY;
```
I guess this is no longer used, right?
```suggestion
```

tomheller commented on libs/barista-components/table/src/table-data-source.ts:
```diff
@@ -463,6 +517,7 @@ export class DtTableDataSource<T> extends DataSource<T> {
   disconnect(): void {
     this._renderChangesSubscription.unsubscribe();
     this._searchChangeSubscription.unsubscribe();
+    this._selectionChangeSubscription.unsubscribe();
```
```suggestion
```

tomheller commented:
Hey @dt-mafe!
It seems there is something weird going on with your commit history. Before we merge this one, we might need to clean up the history a bit.

dt-mafe commented:
> Hey @dt-mafe!
> It seems there is something weird going on with your commit history. Before we merge this one, we might need to clean up the history a bit.
> 
> Please reach out if you need any help with the rebase / squashing of the commits.

Not really sure what happened there. Anyway, I cleaned up the git history a bit and rebased master again and it looks fine again.

However, bazel_stylelint build step is now failing and I don't really know why.
Errors are:
`setup_ci_build_cache.sh: 7: [: pull/1071: unexpected operator
Please specify the "BAZEL_REMOTE_CACHE_PASSWORD" environment variable when setting up remote  cache`
and
`ERROR: Could not open auth credentials file './tools/bazel_build_cache/cache-write-key.json': ./tools/bazel_build_cache/cache-write-key.json (No such file or directory)`
which I would not really suspect to be related to my changes.

tomheller commented:
@dt-mafe Thanks for cleaning up the git history. 

We are aware of the bazel stylelint errors. This is a configuration fault on our end nothing on your end broke this check. 
This is fine and we will review again soon 👍
